# Movie Recommendation Algorithm

As part of this task, you are expected follow the instructions below and create a movie recommendation algorithm to make your users happy :)

### 1. We start off with the known dependencies we need

Pandas is a must. Pandas is a powerful data analysis library built for Python users. It helps us manipulate complicated data in a user-friendly manner. You will understand soon the convenience of it and come to love it as much as we do. 

Please use the run button on the right hand side to execute the code block below. Once the dependencies are imported successfully, you will see a green tick at the bottom of the code block.

In [3]:
import pandas as pd # refer to pandas as pd
from movie_details import get_movies_by_id  

### 2. Then we need to get our movies dataset

The dataset includes thousands of movies and detailed information for each one of them. Let's see how it looks like shall we? 

Run the code block below to execute the code. Once the code block runs, the dataset will be displayed underneath the code block.

In [7]:
path = "./data/dataset2.csv"
df = pd.read_csv(path) # df stands for Data Frame
df.head(10) # uncomment to see how horrible the dataset looks

Unnamed: 0  adult                     backdrop_path    budget  \
0         NaN  False  /dvQj1GBZAZirz1skEEZyWH2ZqQP.jpg         0   
1         NaN  False                               NaN         0   
2         NaN  False  /uJlc4aNPF3Y8yAqahJTKBwgwPVW.jpg         0   
3         NaN  False                               NaN         0   
4         NaN  False                               NaN         0   
5         NaN  False  /hQ4pYsIbP22TMXOUdSfC2mjWrO0.jpg         0   
6         NaN  False  /l94l89eMmFKh7na2a1u5q67VgNx.jpg         0   
7         NaN  False  /c1BaOxC8bo5ACFYkYYxL0bBWRaq.jpg   4000000   
8         NaN  False  /bGMqHn0H2UY6SPZ5Vch4YJM2jDO.jpg  21000000   
9         NaN  False                               NaN     42000   

                                              genres  \
0  [{'id': 35, 'name': 'Comedy'}, {'id': 10751, '...   
1                  [{'id': 12, 'name': 'Adventure'}]   
2  [{'id': 18, 'name': 'Drama'}, {'id': 10749, 'n...   
3                                                 []   
4                  [{'id': 10751, 'name': 'Family'}]   
5  [{'id': 18, 'name': 'Drama'}, {'id': 35, 'name...   
6  [{'id': 18, 'name': 'Drama'}, {'id': 35, 'name...   
7  [{'id': 80, 'name': 'Crime'}, {'id': 35, 'name...   
8  [{'id': 28, 'name': 'Action'}, {'id': 80, 'nam...   
9                [{'id': 99, 'name': 'Documentary'}]   

                                    homepage     id    imdb_id  \
0                                        NaN   3924  tt0029927   
1                                        NaN   6124  tt0011436   
2                                        NaN   8773  tt0055747   
3                    http://www.nwdfilms.com  25449        NaN   
4                                        NaN  31975  tt1656746   
5                                        NaN      2  tt0094675   
6                                        NaN      3  tt0092149   
7  https://www.miramax.com/movie/four-rooms/      5  tt0113101   
8                                        NaN      6  tt0107286   
9                     http://lifeinloops.com      8  tt0825671   

  original_language                      original_title  ...   revenue  \
0                en                             Blondie  ...         0   
1                de                 Der Mann ohne Namen  ...         0   
2                fr                 L'Amour à vingt ans  ...         0   
3                en  New World Disorder 9: Never Enough  ...         0   
4                en      Sesame Street: Elmo Loves You!  ...         0   
5                fi                               Ariel  ...         0   
6                fi                Varjoja paratiisissa  ...         0   
7                en                          Four Rooms  ...   4257354   
8                en                      Judgment Night  ...  12136938   
9                en    Life in Loops (A Megacities RMX)  ...         0   

   runtime                                   spoken_languages    status  \
0       70  [{'english_name': 'English', 'iso_639_1': 'en'...  Released   
1      420                                                 []  Released   
2      110  [{'english_name': 'Italian', 'iso_639_1': 'it'...  Released   
3       69  [{'english_name': 'English', 'iso_639_1': 'en'...  Released   
4       46                                                 []  Released   
5       73  [{'english_name': 'Finnish', 'iso_639_1': 'fi'...  Released   
6       74  [{'english_name': 'English', 'iso_639_1': 'en'...  Released   
7       98  [{'english_name': 'English', 'iso_639_1': 'en'...  Released   
8      109  [{'english_name': 'English', 'iso_639_1': 'en'...  Released   
9       80  [{'english_name': 'English', 'iso_639_1': 'en'...  Released   

                                             tagline  \
0  The favorite comic strip of millions at last o...   
1                                                NaN   
2               The Intimate Secrets of Young Lovers   
3                                            

As you can see, the dataset doesn't look very pretty, does it? That's why we need Pandas to get exactly what we need from the dataset. We need the following three columns from the dataset:
* title
* imdbID
* overview

Try running the code block below to see how to get your desired columns from a dataset.

In [8]:
df[['title', 'imdb_id', 'overview']]    

title    imdb_id  \
0                                 Blondie  tt0029927   
1           Peter Voss, Thief of Millions  tt0011436   
2                          Love at Twenty  tt0055747   
3      New World Disorder 9: Never Enough        NaN   
4          Sesame Street: Elmo Loves You!  tt1656746   
...                                   ...        ...   
1515                  Geliebte weiße Maus  tt0058134   
1516       Attack of the Killer Tomatoes!  tt0080391   
1517  Hindenburg Disaster: Probable Cause  tt0418729   
1518                        Maus und Katz  tt0107532   
1519                          Big Trouble  tt0246464   

                                               overview  
0     Blondie and Dagwood are about to celebrate the...  
1                                                   NaN  
2     Love at Twenty unites five directors from five...  
3     Gee Atherton ripping the Worlds course the day...  
4     Elmo is making a very, very super special surp...  
...                                                 ...  
1515  „White mouse“ Fritz controls the traffic on Dr...  
1516  Attack of the Killer Tomatoes is a 1978 comedy...  
1517                                                NaN  
1518                                                NaN  
1519  The story of how a mysterious suitcase brings ...  

[1520 rows x 3 columns]

Let's create a function that returns the **title**, **imdbID** and **overview** of all the movies in the dataset.

In [6]:
def get_dataset():
    path = "./data/dataset2.csv"
    df = pd.read_csv(path)
    df = df[['title', 'imdb_id', 'overview']]
    df['overview'] = df['overview'].fillna('') # replace NaN values with an empty string
    return df

df = get_dataset() # trigger the function and save it as a Dataframe
df.head(5) # Display the first 5 entries of the Dataframe

title    imdb_id  \
0                             Blondie  tt0029927   
1       Peter Voss, Thief of Millions  tt0011436   
2                      Love at Twenty  tt0055747   
3  New World Disorder 9: Never Enough        NaN   
4      Sesame Street: Elmo Loves You!  tt1656746   

                                            overview  
0  Blondie and Dagwood are about to celebrate the...  
1                                                     
2  Love at Twenty unites five directors from five...  
3  Gee Atherton ripping the Worlds course the day...  
4  Elmo is making a very, very super special surp...

### 3. It's time to process our user's request

Our users have requested a feature where when they add a new movie to their favourites, they would like to get 2 new movie recommendations. The frontend has already been prepared for this task. All there is left to do is:
* Get the movie details from OMDB API using the IMDB ID of the movie
* Add the details to our dataset 
* If it already exists in the dataset, remove duplicates

#### 3.1 We will make use of the new API endpoint we created to get movie details by IMDB ID. 

Run the code block below and test it out!

In [9]:
response = get_movies_by_id("tt0029927") # get movie details with IMDB ID
response = pd.json_normalize(response) # normalize response from OMDB API into a flat table
response

Title  Year   Rated     Released Runtime           Genre  \
0  Blondie  1938  Passed  30 Nov 1938  70 min  Comedy, Family   

           Director                        Writer  \
0  Frank R. Strayer  Richard Flournoy, Chic Young   

                                      Actors  \
0  Penny Singleton, Arthur Lake, Larry Simms   

                                                Plot  ... Metascore  \
0  Dagwood loses his job on the eve of his and Bl...  ...       N/A   

  imdbRating imdbVotes     imdbID   Type  DVD BoxOffice Production Website  \
0        6.9       793  tt0029927  movie  N/A       N/A        N/A     N/A   

  Response  
0     True  

[1 rows x 25 columns]

So it works! But as you can see the column names do not match with the ones that the dataset has for the same information... 

| column info | column name in dataset | column name in API response |
| --- | --- | --- |
|Title |title|Title|
|IMDB ID|imdb_id| imdbID|
|Overview | overview| Plot |

##### 3.2 Let's create a function that adds the neccessary details of the favourited movie to the dataset

Important points to consider:
* Function should take an IMDB ID as an input
* The column names should match
* There shouldn't be any duplicates in the dataset
* Function should return the updated dataframe

In [85]:
# Write your code here    


### 4. Create the recommendation algorithm

We will be focusing on the plot of the movies and use NLP (Natural Language Processing) to find similarities between the plot of the favourited movie and all the other movie plots in our database.

First, we need to import the necessary Python machine learning libraries that we can use to complete the task. 

Scikit-learn (full name for sklearn) is a machine learning library for Python programming. It includes simple and efficient tools for predictive data analysis. [Check here for more](https://scikit-learn.org/stable/index.html).

We will use one of it's functionalities that allows us to perform data analysis on text. We will compute pairwise similarity scores for all movies based on their plot descriptions and recommend movies based on that similarity score.

In [10]:
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import linear_kernel

Now if you are wondering how we will achieve that, one way of doing it is to create a Term Frequency-Inverse Document Frequency (TF-IDF) matrix... 

In human words, **TF (Term Frequency)** is the relative frequency of a word in a document and is given as (term instances/total instances). 

**IDF (Inverse Document Frequency)** is the relative count of documents containing the term is given as log(number of documents/documents with term).

The overall importance of each word to the documents in which they appear is equal to **TF * IDF**.

This will give you a matrix where each column represents a word in the overview vocabulary (all the words that appear in at least one document) and each row represents a movie, as before.This is done to reduce the importance of words that occur frequently in plot overviews and therefore, their significance in computing the final similarity score.

In [9]:
df = add_plot("tt0029927")

# Define a TF-IDF Vectorizer Object. Remove all english stop words such as 'the', 'a'
tfidf = TfidfVectorizer(stop_words='english')

# Construct the required TF-IDF matrix by fitting and transforming the data
tfidf_matrix = tfidf.fit_transform(df['overview'])

# Output the shape of tfidf_matrix
tfidf_matrix.shape

NameError: name 'add_plot' is not defined

We see that over 11,000 different words were used to describe the 1517 movies in our dataset. Now we can go ahead and calculate the cosine similarity score using our matrix and **linear_kernel()** functionality from the Scikit-learn library.

One explanation of a kernel is as follows: 

> a collection of distinct forms of pattern analysis algorithms, using a linear classifier, they solve an existing non-linear problem

So we will use a linear kernel to classify movies as similar or not.

In [88]:
# Similarity scores between all movies in the dataset
sim = linear_kernel(tfidf_matrix, tfidf_matrix)

# We create a new Series where the index is the IMDB IDs and the values are the actual indices from the original dataframe
indices = pd.Series(df.index, index=df['imdb_id'])

# Let's try getting the index of one movie with its IMDB ID
idx = indices["tt0029927"]

# Get the pairwsie similarity scores of all movies with that movie
sim_scores = list(enumerate(sim[idx]))

# Sort the movies based on the similarity scores
sim_scores = sorted(sim_scores, key=lambda x: x[1], reverse=True)

# Get the scores of the 2 most similar movies
sim_scores = sim_scores[1:3]

# Get the movie indices
movie_indices = [i[0] for i in sim_scores]


#### Final Challenge

Create a function that takes an IMDB ID as an input and returns the IMDB IDs of the recommended two movies and display the output.

In [89]:
# Write your code here


[1304, 296]


['tt0044357', 'tt0428430']

Now let's move all the functions you created to the recommend.py file. Functions to move:
* get_dataset()
* add_plot()
* get_recommendation()